### ISSUE TITLE
Abstract: future

### ISSUE THREAD

**Issue by jbcaillau** [control-toolbox/CTBase.jl#71]
@ocots @PierreMartinon @joseph-gergaud @j-l-s Some todo's suggestions:
- add single sided inequalities (see also #67 as having a kwargs version of `constraint!` does help for this)
- add reverse inequalities (≥)
- declare time, state, control, variable simultaneously according to
```julia
(t, x, u) ∈ [ 0, 1 ] × R³ × R²
(t, x, u, v) ∈ [ 0, v₂ ] × R³ × R² × R²            # meaning that final time is free and equal to v[2] = v₂
```
- `in place` dynamics (`f!(ẋ(t), x(t), u(t))`) or even DAE form (`F(ẋ(t), x(t), u(t)) == 0`)
- for the `in place` form, sth like
```julia
ẋ(t) == f(x(t), u(t)), inplace
```
could do
- parse dynamics in several steps: `x'[2](t) == ..., x'[2:4](t) == ...` (each time call `dynamics!`, and assemble afterwards what the dynamics is...)

**Comment by ocots** [comment]
To declare the time, state and control, I propose:

```julia
@def begin
  t ∈ [ 0, 1 ], time
  x ∈ R³, state
  u ∈ R², control
end

# or

@def begin
  (t, x, u) ∈ [ 0, 1 ] × R³ × R²
end
```

```julia
@def begin
  v ∈ R², variable
  t ∈ [ 0, v₂ ], time
  x ∈ R³, state
  u ∈ R², control
end

# or

@def begin
  (t, x, u, v) ∈ [ 0, v₂ ] × R³ × R² × R²
end
```

```julia
@def begin
  t ∈ [ 0, 1 ], time
  x = (a, b, c) ∈ R³, state
  u ∈ R², control
  -1 ≤ u₁ ≤ 1
  0 ≤ c ≤ 1
end

# or

@def begin
  t ∈ [ 0, 1 ], time
  x = (a, b, c) ∈ R² × [ 0, 1 ] , state
  u ∈ [ -1, 1 ] × R, control
end

# or

@def begin
  t ∈ [ 0, 1 ], time
  x = (a, b, c) ∈ R × R × [ 0, 1 ] , state
  u ∈ [ -1, 1 ] × R, control
end
```

**Comment by PierreMartinon** [comment]
For the record, the abstract form is great, I am increasingly using it instead of the 'functional' form :-) It feels a bit like when we added the proper gui for bocop: after a while I very rarely used the command-line since the gui was just better.

The main thing holding me back are the non-standard characters. I'll try to find the time to add some bits (looking at you, dear <=) !

**Comment by ocots** [comment]
You can write non unicode if you want. Check the tutorial 😁

**Comment by jbcaillau** [comment]
@PierreMartinon yes. I believe the functional syntax should probably be kept internal. the only advantage is that it is incremental (and that you can delete a constraint...) but re-defining an abstract problem just costs nothing so...

as mentioned by @ocots, full non-unicode support now. sone that wasn't you asked so 🙂...

**Comment by jbcaillau** [comment]
related to https://github.com/control-toolbox/CTBase.jl/issues/257
